rule/rule.go

```diff
@@ -118,6 +147,34 @@ func LoadWorkspaceData(path, pkg string, data []byte) (*File, error) {
 	return ScanAST(pkg, ast), nil
 }
 
+// LoadMacroData parses a bzl file from a byte slice and scans for the load
+// statements and the rules called from the given Starlark function. If there is
+// no matching function name, then a new function with that name will be created.
+// The function's syntax tree will be returned within File and can be modified by
+// Sync and Save calls.
+func LoadMacroData(path, pkg, defName string, data []byte, isEmpty bool) (*File, error) {
+	ast, err := bzl.ParseDefault(path, data)
+	if err != nil {
+		return nil, err
+	}
+	f := &File{
+		File: ast,
+		Pkg:  pkg,
+		Path: ast.Path,
+	}
+	if !isEmpty {
+		_, f.Loads, f.function = ScanASTBody(defName, ast.Stmt)
+	}
+	if f.function == nil {
+		defStmt := bzl.DefStmt{Name: defName}
+		ast.Stmt = append(ast.Stmt, &defStmt)
```

Comment:
Let's not add the function until `Sync`. We'll need a way to track whether it's part of the AST in case `Sync` is called multiple times.

Currently, if we don't end up adding any rules to the function, we'll have an empty body, and the formatter will complain. `Sync` should add a `pass` statement if the body would be empty.

We should also be able to replace a `pass` statement if the function already exists and that's the only thing there.